codex/erasure/erasure.nim

```diff
@@ -269,14 +291,88 @@ proc init*(
     strategy: strategy,
   )
 
+proc leopardEncodeTask(tp: Taskpool, task: ptr EncodeTask) {.gcsafe.} =
+  # Task suitable for running in taskpools - look, no GC!
+  let encoder =
+    task[].erasure.encoderProvider(task[].blockSize, task[].blocksLen, task[].parityLen)
+  defer:
+    encoder.release()
+    discard task[].signal.fireSync()
+
+  if (
+    let res =
+      encoder.encode(task[].blocks, task[].parity, task[].blocksLen, task[].parityLen)
+    res.isErr
+  ):
+    warn "Error from leopard encoder backend!", error = $res.error
+
+    task[].success.store(false)
+  else:
+    task[].success.store(true)
+
+proc encodeAsync*(
+    self: Erasure,
+    blockSize, blocksLen, parityLen: int,
+    data: ref seq[seq[byte]],
+    parity: ptr UncheckedArray[ptr UncheckedArray[byte]],
+): Future[?!void] {.async: (raises: [CancelledError]).} =
+  without threadPtr =? ThreadSignalPtr.new():
+    return failure("Unable to create thread signal")
+
+  defer:
+    threadPtr.close().expect("closing once works")
+
+  var blockData = createDoubleArray(blocksLen, blockSize)
+
+  for i in 0 ..< data[].len:
+    copyMem(blockData[i], addr data[i][0], blockSize)
+
+  defer:
+    freeDoubleArray(blockData, blocksLen)
+
+  ## Create an ecode task with block data 
+  var task = EncodeTask(
+    erasure: addr self,
+    blockSize: blockSize,
+    blocksLen: blocksLen,
+    parityLen: parityLen,
+    blocks: blockData,
+    parity: parity,
+    signal: threadPtr,
+  )
+
+  let t = addr task
+
+  doAssert self.taskPool.numThreads > 1,
+    "Must have at least one separate thread or signal will never be fired"
+  self.taskPool.spawn leopardEncodeTask(self.taskPool, t)
+  let threadFut = threadPtr.wait()
+
+  try:
+    await threadFut.join()
+  except CatchableError as exc:
+    try:
+      await threadFut
+    except AsyncError as asyncExc:
```

Comment:
Why AsyncError here and not CatchableError?

Reply:
@dryajov AsyncError would be more appropriate as threadPtr.wait() raises an AsyncError, [ref](https://github.com/status-im/nim-chronos/blob/0646c444fce7c7ed08ef6f2c9a7abfd172ffe655/chronos/threadsync.nim#L367)